### PR TITLE
Update _merge_token_usage() to canonical field names with old-name fallback

### PR DESCRIPTION
**Pull request by Trecek**
## Summary

Replace the hardcoded four-key tuple in `_merge_token_usage()` with a canonical-key-first iteration that falls back to legacy Anthropic API field names. The function will read from either naming convention but always write merged values under canonical keys. No `CanonicalTokenUsage` import — stays dict-based.

### New (★):
tests/execution/test_merge_token_usage.py

### Modified (●):
src/autoskillit/execution/headless/_headless_recovery.py
tests/execution/CLAUDE.md

Closes #2455

## Implementation Plan

Plan file: `/home/talon/projects/autoskillit-runs/impl-20260517-144156-049292/.autoskillit/temp/make-plan/p2-a11-wp1-update-merge-token-usage-canonical-fields_plan_2026-05-17_144500.md`

🤖 Generated with [Claude Code](https://claude.com/claude-code) via AutoSkillit

## Token Usage Summary

| Step | Model | count | uncached | output | cache_read | peak_ctx | turns | cache_write | time |
|------|-------|-------|----------|--------|------------|----------|-------|-------------|------|
| plan | claude-sonnet-4-6 | 1 | 70 | 6.6k | 633.5k | 63.9k | 77 | 42.6k | 4m 52s |
| verify | claude-sonnet-4-6 | 1 | 41 | 8.8k | 534.5k | 52.6k | 103 | 35.2k | 6m 28s |
| implement* | MiniMax-M2.7-highspeed | 1 | 364.4k | 4.5k | 650.8k | 46.5k | 46 | 56.7k | 2m 8s |
| prepare_pr* | MiniMax-M2.7-highspeed | 1 | 72.2k | 2.9k | 209.6k | 29.9k | 19 | 42.6k | 1m 14s |
| compose_pr* | MiniMax-M2.7-highspeed | 1 | 45.5k | 1.5k | 149.7k | 29.9k | 15 | 42.5k | 44s |
| review_pr | claude-sonnet-4-6 | 3 | 1.6k | 67.6k | 1.5M | 68.2k | 115 | 148.6k | 16m 44s |
| resolve_review | claude-sonnet-4-6 | 3 | 761 | 78.2k | 5.8M | 117.2k | 227 | 225.3k | 37m 49s |
| **Total** | | | 484.6k | 170.0k | 9.5M | 117.2k | | 593.4k | 1h 10m |

\* *Step used a non-Anthropic provider; caching behavior may differ.*

## Token Efficiency

| Step | LoC Changed | cache_read/LoC | cache_write/LoC | output/LoC |
|------|-------------|----------------|-----------------|------------|
| plan | 0 | — | — | — |
| verify | 0 | — | — | — |
| implement | 145 | 4488.2 | 391.1 | 30.9 |
| prepare_pr | 0 | — | — | — |
| compose_pr | 0 | — | — | — |
| review_pr | 0 | — | — | — |
| resolve_review | 55 | 104810.9 | 4095.7 | 1421.6 |
| **Total** | **200** | 47357.6 | 2967.2 | 850.2 |
## Model Usage Breakdown

| Model | steps | uncached | output | cache_read | cache_write | time |
|-------|-------|----------|--------|------------|-------------|------|
| claude-sonnet-4-6 | 2 | 1.1k | 94.0k | 9.3M | 275.3k | 44m 21s |
| claude-opus-4-6 | 2 | 1.1k | 98.8k | 7.9M | 462.7k | 57m 59s |
| MiniMax-M2.7-highspeed | 4 | 3.1M | 51.1k | 6.5M | 453.4k | 34m 33s |

### CODE REVIEW

File: tests/execution/test_merge_token_usage.py [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+        result = _merge_token_usage(None, nudge)
+        assert result is nudge
+        assert "cache_creation_input_tokens" in result
+        assert "cache_read_input_tokens" in result
+
+    def test_both_none_returns_none(self):
+        """Both None → return None."""
+        assert _merge_token_usage(None, None) is None
+
+    def test_non_numeric_values_skipped(self):
```
Comment (Trecek) [inline tests/execution/test_merge_token_usage.py:128]: [warning] tests: The production code silently drops valid nudge data when base contains a non-numeric value for the same key (isinstance check fails, so merged[canonical] is never written, leaving the non-numeric base value intact). The test blesses this as the contract at L121. If this is an accepted limitation, the production code should document it; if not, the test is normalizing a latent bug. Clarify intent.

File: src/autoskillit/execution/headless/_headless_recovery.py [hunk 2]
```diff
@@ -323,14 +323,16 @@ def _merge_token_usage(
...
-        "output_tokens",
-        "cache_creation_input_tokens",
-        "cache_read_input_tokens",
-    ):
-        base_val = base.get(key, 0)
-        nudge_val = nudge.get(key, 0)
-        if isinstance(base_val, (int, float)) and isinstance(nudge_val, (int, float)):
-            merged[key] = base_val + nudge_val
+    for canonical, legacy in _CANONICAL_TO_LEGACY.items():
+        b = base.get(canonical) if canonical in base else base.get(legacy) if legacy else None
```
Comment (Trecek) [inline src/autoskillit/execution/headless/_headless_recovery.py:327]: [warning] cohesion: If base[canonical] is explicitly set to None (e.g., base={"input_tokens": None}), then `canonical in base` is True so the ternary assigns b=None; no fallback to the legacy key occurs. If nudge also has canonical=None or missing, b=None and n=None → continue skips the field entirely, even if a legacy alias existed. This edge case is not covered by any test and silently drops data.

File: src/autoskillit/execution/headless/_headless_recovery.py [hunk 3]
```diff
@@ -323,14 +323,16 @@ def _merge_token_usage(
...
+        n = nudge.get(canonical) if canonical in nudge else nudge.get(legacy) if legacy else None
+        if b is None and n is None:
+            continue
+        bv = b if b is not None else 0
+        nv = n if n is not None else 0
+        if isinstance(bv, (int, float)) and isinstance(nv, (int, float)):
+            merged[canonical] = bv + nv
```
Comment (Trecek) [inline src/autoskillit/execution/headless/_headless_recovery.py:334]: [warning] cohesion: When base contains both canonical and legacy keys for the same field simultaneously (e.g., base={"cache_write_tokens": 10, "cache_creation_input_tokens": 5}), the merge loop reads the canonical value (10), writes merged[canonical], and the cleanup loop later deletes the legacy key. The legacy value (5) is silently lost. No test covers this coexistence scenario to confirm it is the intended behavior.

**Review by Trecek — COMMENTED** [review]
AutoSkillit PR Review — Verdict: approved_with_comments

**Review by Trecek — COMMENTED** [review]
AutoSkillit review: warning-only findings detected. See inline comments — no blocking changes required.

File: tests/execution/test_merge_token_usage.py (hunk 1 above)
Comment (Trecek) [reply tests/execution/test_merge_token_usage.py:128]: Investigated — this is intentional. The test comment in commit 03289d30 documents the contract: '# base_val fails isinstance check → nudge's 200 is dropped; base value survives unchanged'. The production code correctly preserves the non-numeric base value unchanged per the defensive-skip design (CLAUDE.md's no-comment policy makes the test the appropriate documentation site). No production code change is needed.

File: src/autoskillit/execution/headless/_headless_recovery.py (hunk 2 above)
Comment (Trecek) [reply src/autoskillit/execution/headless/_headless_recovery.py:327]: Investigated — this is intentional. Token usage dicts from Claude's API always have numeric values or are absent entirely; explicit None values for canonical token keys are not a production scenario. Commit 6d7a198b confirms canonical-takes-precedence as the design intent (with explicit debug logging for coexistence). The b is None and n is None → continue guard correctly handles the 'both absent' case.

File: src/autoskillit/execution/headless/_headless_recovery.py (hunk 3 above)
Comment (Trecek) [reply src/autoskillit/execution/headless/_headless_recovery.py:334]: Agreed — fixed in 00eff9a9. Added test_canonical_wins_when_both_canonical_and_legacy_coexist to document that canonical takes precedence when both keys coexist (legacy value is intentionally discarded). The intended behavior was established in commit 6d7a198b with an explicit debug log statement; the new test makes this contract visible and machine-checkable.
